src/masternodes/mn_checks.cpp

```diff
@@ -1029,6 +1115,422 @@ class CCustomTxApplyVisitor : public CCustomTxVisitor
+        if (order.orderType != CICXOrder::TYPE_INTERNAL && order.orderType != CICXOrder::TYPE_EXTERNAL)
+            return Res::Err("invalid order type!");
+
+        if (order.orderType == CICXOrder::TYPE_INTERNAL)
+        {
+            auto tokenFrom = mnview.GetToken(order.idToken);
+            if (!tokenFrom)
+                return Res::Err("%s: %s", __func__, "tokenFrom (" +  tokenFrom->CreateSymbolKey(order.idToken) + ") does not exist!");
+
+            if (order.ownerAddress.empty())
+                return Res::Err("ownerAddress must be not-null!");
+
+            // subtract the balance from tokenFrom to dedicate them for the order
+            CScript txidAddr(order.creationTx.begin(), order.creationTx.end());
+            res = ModifyTokenBalance(order.idToken, order.amountFrom, order.ownerAddress, txidAddr);
+            if (!res.ok)
+                return Res::Err("%s: %s", __func__, res.msg);
+        }
+        else if (order.orderType == CICXOrder::TYPE_EXTERNAL)
+        {
+            auto tokenTo = mnview.GetToken(order.idToken);
+            if (!tokenTo)
+                return Res::Err("%s: %s", __func__, "tokenTo id (" +  tokenTo->CreateSymbolKey(order.idToken) + ") does not exist!");
+        }
+
+        if (order.amountFrom == 0)
+            return Res::Err("order amountFrom must be greater than 0!");
+        if (order.amountToFill != order.amountFrom)
+            return Res::Err("order amountToFill does not equal to amountFrom!");
+        if (order.orderPrice == 0)
+            return Res::Err("order price must be greater than 0!");
+        if (order.expiry == 0)
+            return Res::Err("order expiry must be greater than 0!");
+
+        return mnview.ICXCreateOrder(order);
+    }
+
+    Res operator()(const CICXMakeOfferMessage& obj) const {
+        auto res = CheckICXTx();
+        if (!res)
+            return res;
+
+        CICXMakeOfferImplemetation makeoffer;
+        static_cast<CICXMakeOffer&>(makeoffer) = obj;
+
+        makeoffer.creationTx = tx.GetHash();
+        makeoffer.creationHeight = height;
+
+        auto order=mnview.GetICXOrderByCreationTx(makeoffer.orderTx);
+        if (!order)
+            return Res::Err("%s: %s", __func__, "order with creation tx " + makeoffer.orderTx.GetHex() + " does not exists!");
+        
+        CAmount calcedAmount(static_cast<CAmount>((arith_uint256(order->amountToFill) * arith_uint256(order->orderPrice) / arith_uint256(COIN)).GetLow64()));
+        if (calcedAmount < makeoffer.amount)
+            return Res::Err("%s: %s", __func__, "cannot fill order with that amount, order (" + order->creationTx.GetHex() + ") has less amount to fill!");
+
+        CScript txidAddr(makeoffer.creationTx.begin(),makeoffer.creationTx.end());
+        
+        //calculating DFI per BTC
+        auto DFIBTCpair = mnview.GetPoolPair({5});
+        CAmount DFIperBTC = (arith_uint256(DFIBTCpair->reserveB) * arith_uint256(COIN) / DFIBTCpair->reserveA).GetLow64();
```

Comment:
Get hardcoded poolpair 5? It crashes the node on my regtest, and the `DFIBTCpair` should be checked first?

Reply:
Yeah, I haven't tested this yet. Just did the implementation first. I need to check for mainnet, testnet, devnet and regtest because pool pairs are different and on some they do not exist!